DISCUSSION-23: Allow to chose the storage location by "application"

surli opened the pull request:
## JIRA
https://jira.xwiki.org/browse/DISCUSSION-23

## Description
This commit starts a big refactoring of Discussion to integrate the
concept of applicationHint: the idea is to be able to manipulate
discussion with a dedicated applicationHint so that dedicated
applications can be built on top of it. Right now this hint is only used
for choosing the location where to store the discussions, but we could
imagine other usecases later with this hint.

surli committed:
DISCUSSION-23: Allow to chose the storage location by "application"

This commit starts a big refactoring of Discussion to integrate the
concept of applicationHint: the idea is to be able to manipulate
discussion with a dedicated applicationHint so that dedicated
applications can be built on top of it. Right now this hint is only used
for choosing the location where to store the discussions, but we could
imagine other usecases later with this hint.

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionContextService.java:
```diff
@@ -55,12 +58,11 @@
      */
-    Optional<DiscussionContext> getOrCreate(String name, String description, String referenceType,
-        String entityReference);
+    Optional<DiscussionContext> getOrCreate(String applicationHint, String name, String description,
```
Missing javadoc for applicationHint

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:
```diff
@@ -0,0 +1,27 @@
+
+import org.xwiki.contrib.discussions.domain.references.AbstractDiscussionReference;
+
+public interface DiscussionReferencesResolver
```
Missing javadoc

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:
```diff
@@ -0,0 +1,47 @@
+     * type. Note that the serialization might not contain any information about the actual type, so it might not be
+     * possible to check that the type is correct for this reference.
+     *
+     * @param serializedReference a serialization of an {@link AbstractDiscussionReference}.
```
Shouldn't finish with a dot

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesSerializer.java:
```diff
@@ -0,0 +1,43 @@
+     * @param discussionReference the reference to serialize.
+     * @return a serialization of the reference.
+     */
+    String serialize(AbstractDiscussionReference discussionReference);
```
Why not one method per kind of reference? Unless we plan to make it open to others to provide their own reference kinds?

surli replied:
So for the serialization it's actually not needed, you can serialize all the `AbstractDiscussionReference` the same way, so I don't see the point of adding complexity to chose the right method for serializing.

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:
```diff
@@ -0,0 +1,47 @@
+     * @param <T> the actual given concrete type
+     * @return an instance of the given type with the information of the reference.
+     */
+    <T extends AbstractDiscussionReference> T resolve(String serializedReference, Class<T> type);
```
Might be interesting to add resolveXXX method, one for each kind of serialized reference kind. Could be useful when we already know statically the kind of the reference to resolve.

surli replied:
Personally I think it's enough to have a single resolve method taking the type.

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionsActorServiceResolver.java:
```diff
@@ -40,5 +40,5 @@
      * @return the actor service, {@link Optional#empty()} if the resolution failed
      */
-    Optional<DiscussionsActorService> get(String type);
+    DiscussionsActorService get(String type);
```
(small detail) Why is the Optional removed? I agree the default implementation does not return the empty value, but could be the case for another implementation.

surli replied:
So I improved a bit the javadoc there to specify that the contract of this resolver should always fallback on a default service if a specific implementation for the type does not exist. So it should never be the case to return null on another implementation.

manuelleduc commented on discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:
```diff
@@ -0,0 +1,47 @@
+     * @param <T> the actual given concrete type
+     * @return an instance of the given type with the information of the reference.
+     */
+    <T extends AbstractDiscussionReference> T resolve(String serializedReference, Class<T> type);
```
Why not Optional<T> as the resolution could fail?

surli replied:
The resolution could only fail if someone provides a new kind of AbstractDiscussionReference and don't provide an appropriate constructor. So in practice it should never happen, and it wouldn't be the responsibility of the API consumer to be aware of that. For the consumer of the API the reference should always be resolved.

manuelleduc commented on discussions-default/src/main/java/org/xwiki/contrib/discussions/internal/DefaultDiscussionReferencesSerializer.java:
```diff
@@ -0,0 +1,52 @@
+    {
+        String result = discussionReference.getReference();
+        if (!StringUtils.isEmpty(discussionReference.getApplicationHint())) {
+            result = String.format("%s;applicationHint=%s", result, discussionReference.getApplicationHint());
```
We need to discuss what's the best serialization format.

surli replied:
So as detailed in the javadoc, the format is based on XWiki Page reference format which accepts parameters key/value, see: https://extensions.xwiki.org/xwiki/bin/view/Extension/Model%20Module#HPageReferences

manuelleduc commented on discussions-default/src/main/java/org/xwiki/contrib/discussions/internal/DefaultDiscussionsRightService.java:
```diff
@@ -84,6 +84,7 @@ public void initialize() throws InitializationException
         }
     }
 
+
```
useless change ;)

manuelleduc commented on discussions-rest/discussions-rest-default/src/checkstyle/checkstyle-suppressions.xml:
```diff
@@ -26,4 +26,5 @@
 
 <suppressions>
   <suppress checks="ParameterNumber" files="DefaultDiscussionREST.java"/>
+  <suppress checks="ClassFanOutComplexity" files="DefaultDiscussionREST.java"/>
```
We'll need to work on that on a second time :)

manuelleduc reviewed (commented):
Partial review, mainly notes for myself.

surli committed:
DISCUSSION-23: Allow to chose the storage location by "application"

  * Fix tests and coverage
  * Refactor package names in messagestream
  * Fix some bugs in REST API
  * Start fixing UI
  * Add converters so that the script service can work when references
    are used

surli committed:
DISCUSSION-23: Allow to chose the storage location by "application"

  * Fix remaining issues for using messagestream

surli committed:
DISCUSSION-23: Allow to chose the storage location by "application"

  * Get rid of discussion-test-docker module since the discussion
    application should only be tested with docker on a concrete
implementation such as with message stream.

surli committed:
DISCUSSION-23: Allow to chose the storage location by "application"

  * javadoc typo

surli committed:
DISCUSSION-23: Allow to chose the storage location by "application"

  * Refactor to add ActorReference type
